### ISSUE TITLE
Update to Generic Port Interface

### ISSUE THREAD

**Issue by bradjc** [lab11/signpost-software#10]
Before the interface is cemented:

### More specific functions
```c
port_signpost_gpio_enable_interrupt
port_signpost_gpio_disable_interrupt
```

should be

```c 
port_signpost_mod_in_enable_interrupt
port_signpost_mod_in_disable_interrupt
```
to match the other `mod` gpio functions.

### Go back to `printf`
Is there a reason to not use the industry standard `printf()` function? This seems like a very custom interface that we don't really need:

```
#define PORT_PRINT_MAX_LEN 80
extern char port_print_buf[PORT_PRINT_MAX_LEN];
void port_signpost_debug_print(char * msg);
```

If a platform wants to use a fixed length buffer and `snprintf` then it should do that inside of its own implementation.

**Comment by adkinsjd** [comment]
The primary problem is that arduino does not have printf.

I think to solve that problem we should define printf in the port layer and force people on a platform like arduino where printf isn't implemented to implement it.

**Comment by adkinsjd** [comment]
@nealjack any thoughts on this?

**Comment by ppannuto** [comment]
fwiw, following a discussion with Branden the other day where he was worried this would be burdensome, writing a printf compatible wrapper is quite short:

```c
#include <stdarg.h>
#include <stdio.h>
#include <string.h>

__attribute__ ((format (printf, 1, 2)))
int port_printf(char* fmt, ...) {
	va_list params;
	va_start(params, fmt);

	char buf[256];

	int ret = vsnprintf(buf, 256, fmt, params);
	if (ret < 0) {
		return ret;
	}
	if (ret > 256) {
		const char* m = "Message Truncated:";
		memcpy(buf, m, strlen(m));
		ret = 256;
	}
	
	// Serial.write(buf) or whatever
	
	va_end(params);
	return ret;
}

int main() {
	port_printf("%d\n", 1);
	return 0;
}
```

**Comment by nealjack** [comment]
Yeah I don't see any point in keeping the old print interface if a printf wrapper is that easy.

**Comment by adkinsjd** [comment]
Is it even necessary to make it port_printf? Is there any easy way just to direct printf to our redefined printf? (I really have no idea how this would work)

**Comment by ppannuto** [comment]
Not sure to be honest. I suspect it's fine to just use printf - on
platforms (like Tock) that support it it'll just work. On platforms like
Ardunio that don't, using it would normally result in a linker error, but
not if we define our own function.

I think it only gets funny if Arduino or something has a stub printf
implementation what we have to override somehow

**Comment by ppannuto** [comment]
(where "we" is the port author)

**Comment by prabaldutta** [comment]
One thing to consider is whether printf is blocking or not, and if not,
what the reentrancy issues, and what behavior is expected when multiple
callers invoke (long) writes concurrently.

- Prabal

**Comment by adkinsjd** [comment]
@ppannuto so would that mean printf only gets defined and implemented on platforms like arduino, and we comment it out otherwise? that's a bit strange

**Comment by ppannuto** [comment]
No, that would mean that `printf` is implemented by
`arudino_signbus_port.c`, and if mbed needed to define a printf as well, it
would just copy this rough implementation to `mbed_signbus_port.c`

**Comment by adkinsjd** [comment]
Yeah I understand that - and then on tock we would not define a printf function because it's already there, but does that mean we would need to comment out the printf function from port_signpost.h? when using it on tock? or would everything still compile even with redefinition of printf in the header file?

**Comment by ppannuto** [comment]
I don't think we need the printf declaration in the port header file at
all. Simply #include <stdio.h> should do the right thing in all cases
